src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,471 @@
+     */
+    @Override
+    protected byte[] engineDeriveData(KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+
+        // inspect KDFParameterSpec object
+        inspectKDFParameterSpec(kdfParameterSpec);
+
+        try {
+            // set up the HMAC instance
+            hmacLen = setupHMAC(hmacAlgName);
+        } catch (NoSuchAlgorithmException nsae) {
+            throw new ProviderException(nsae);
+        }
+
+        if (HKDFTYPE == HKDFTYPES.EXTRACT) {
+            // perform extract
+            try {
+                return hkdfExtract(initialKeyMaterial, (salt == null) ? null : salt.getEncoded());
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF Extract could not be initialized with the given key or salt material");
+            }
+
+        } else if (HKDFTYPE == HKDFTYPES.EXPAND) {
+            // perform expand
+            try {
+                return Arrays.copyOf(hkdfExpand(this.pseudoRandomKey, this.info, this.length),
+                                     this.length);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF Expand could not be initialized with the given key material");
+            }
+
+        } else if (HKDFTYPE == HKDFTYPES.EXTRACTEXPAND) {
+            // perform extract and then expand
+            try {
+                byte[] extractResult = hkdfExtract(initialKeyMaterial,
+                                                   (salt == null) ? null : salt.getEncoded());
+                this.pseudoRandomKey = new SecretKeySpec(extractResult, "RAW");
+                return Arrays.copyOf(hkdfExpand(this.pseudoRandomKey, this.info, this.length),
+                                     this.length);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF ExtractExpand could not be initialized with the given key or salt "
+                    + "material");
+            }
+        }
+
+        return null;
+    }
+
+    protected int setupHMAC(String hmacAlgName) throws NoSuchAlgorithmException {
+        hmacObj = Mac.getInstance(hmacAlgName);
+        return hmacObj.getMacLength();
+    }
+
+    protected void inspectKDFParameterSpec(KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+        // A switch would be nicer, but we may need to backport this before JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable in instanceof either
+        if (kdfParameterSpec instanceof HKDFParameterSpec.Extract) {
+            HKDFParameterSpec.Extract anExtract = (HKDFParameterSpec.Extract) kdfParameterSpec;
+            this.ikms = anExtract.ikms();
+            this.salts = anExtract.salts();
+            if (isNullOrEmpty(ikms) && isNullOrEmpty(salts)) {
```

Comment:
As I said earlier, I think this restriction is not necessary.

Reply:
Yes, it is not possible for the params to be null as `HKDFParameterSpec.Extract` is final and always passes in non-null lists to them.